src/lightning/pytorch/core/mixins/hparams_mixin.py

```diff
@@ -111,19 +111,19 @@ class ``__init__`` to be ignored
                 frame = current_frame.f_back
         save_hyperparameters(self, *args, ignore=ignore, frame=frame)
 
-    def _set_hparams(self, hp: Union[MutableMapping, Namespace, str]) -> None:
+    def _set_hparams(self, hp: Union[MutableMapping, Namespace]) -> None:
         hp = self._to_hparams_dict(hp)
 
         if isinstance(hp, dict) and isinstance(self.hparams, dict):
             self.hparams.update(hp)
         else:
-            self._hparams = hp
+            self._hparams: AttributeDict = hp
 
     @staticmethod
-    def _to_hparams_dict(hp: Union[MutableMapping, Namespace, str]) -> Union[MutableMapping, AttributeDict]:
+    def _to_hparams_dict(hp: Union[MutableMapping, Namespace]) -> AttributeDict:
         if isinstance(hp, Namespace):
             hp = vars(hp)
-        if isinstance(hp, dict):
+        if isinstance(hp, MutableMapping):
             hp = AttributeDict(hp)
         elif isinstance(hp, _PRIMITIVE_TYPES):
```

Comment:
This `elif` and the one below should not be reached based on the input types to `_to_hparams_dict` in the first place. I think this should rather just be an `else: raise ValueError("unsupported types ...")`

Reply:
The code looks like it wants to just adjust the error message a bit to make it more specific depending on what the wrong type is the user passed in. It's clearer if you write the two elif's as:

```
else:
    if ...
        raise 
    if ...
         raise
```
         
But as you say, there is a missing case for any other type that is not allowed 😄

Reply:
Ah, I misread the second `elif`. Actually, this should work as-is then (though I find it quite convoluted)